### PR TITLE
dependencies: track untracked implied dependencies, wrapup dashboard.

### LINKED ISSUE #12673: Enhance external dependency documentation
https://github.com/envoyproxy/envoy/pull/12639 introduced a dashboard documenting our external dependencies and their version/CPE. There's a number of followup improvements that this issue tracks:

- [x] Add project description field that is human readable.
- [x] Improve handling of version strings that require mangling to different forms, see https://github.com/envoyproxy/envoy/pull/12639#issuecomment-676606571.
- [ ] Split out metadata and join with the labels produced by `bazel query //external:all --output`
- [x] Link GH dependencies directly to the GH tree at the correct tag/SHA.
- [x] Include `api/bazel/repository_locations.py`
- [ ] Include `bazel/dependency_imports.bzl`, possibly UDPA deps.
- [x] Track "last updated" status.
- [x] Distinguish core vs. extension dependencies. Ideally this is done using `bazel query` smarts, so that we can get fine grained and live information, e.g. which extensions are using which deps over time.
- [x] Include version in CPE links
- [x] Update CPE metadata where missing.
- [x] Update `use_category` where incorrect, there are some missing dataplane and controlplane tags on some deps.

### PR DESCRIPTION
**Pull request by htuch**
A few changes that wrapup #12673.

* Python/Go dependencies that aren't part of the Envoy binary build
  don't make sense to track in repository_locations.bzl, since they
  have their own language specific metadata (e.g. requirements.txt)
  or are in many cases transitively implied.

* Ensure that the full set of dependencies visible to bazel query
  is now validated. This requires that we explicitly call out
  transitive dependencies that are implied by direct dependencies
  in repository_locations.bzl. A new annotation `implied_untracked_deps`
  is used.

Fixes #12673

Risk level: Low
Testing: validate.py.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: api/bazel/external_deps.bzl [hunk 1]
```diff
@@ -2,19 +2,38 @@ load("@envoy_api//bazel:repository_locations_utils.bzl", "load_repository_locati
...
+
+    # A list of extensions when 'use_category' contains 'dataplane_ext' or 'observability_ext'.
+    "extensions",
+
+    # Additional dependencies loaded transitively via this dependency that are not tracked in
+    # Envoy (see the external dependency at the given version for information).
+    "implied_untracked_deps",
+
+    # When the dependency was last updated in Envoy.
+    "last_updated",
```
Comment (moderation) [inline api/bazel/external_deps.bzl:21]: As per chat in Slack, I'd like to rename `last_updated` to `release_date`. The comment on line 20 should be something like:
`# When the dependency was updated in its repository. For dependencies that have releases, this is the date of the release. For dependencies without releases or for scenarios where we temporarily need to use a commit, this date should be the date of the commit`

Comment (moderation) [reply api/bazel/external_deps.bzl:21]: https://github.com/envoyproxy/envoy/blob/master/DEPENDENCY_POLICY.md#L48 should be updated with similar language

File: tools/dependency/validate.py [hunk 2]
```diff
@@ -188,7 +203,8 @@ def ValidateControlPlaneDeps(self):
     # these paths.
     queried_controlplane_core_min_deps = self._build_graph.QueryExternalDeps(
         '//source/common/config/...')
-    expected_controlplane_core_deps = self._dep_info.DepsByUseCategory('controlplane')
+    expected_controlplane_core_deps = self._dep_info.DepsByUseCategory('controlplane').union(
```
Comment (asraa) [inline tools/dependency/validate.py:206]: just to clarify: it seems expected controlplane deps will be covered by controlplane and api so it's not really a problem they're entangled like dataplane?

File: tools/dependency/validate.py [hunk 3]
```diff
@@ -91,12 +100,16 @@ def QueryExternalDeps(self, *targets):
...
     for d in deps:
       match = BAZEL_QUERY_EXTERNAL_DEP_RE.match(d)
       if match:
         ext_dep = match.group(1)
-        if ext_dep not in IGNORE_DEPS:
-          ext_deps.add(ext_dep)
+        if ext_dep in self._ignore_deps:
+          continue
+        if ext_dep in self._implied_untracked_deps_revmap:
+          ext_dep = self._implied_untracked_deps_revmap[ext_dep]
```
Comment (asraa) [inline tools/dependency/validate.py:111]: for some reason this was hard for me to gather. could you add a small comment 
"if the dependency is untracked, add the source dependency that loaded it transitively"

**Review by asraa — COMMENTED** [review]
Thanks! I personally think last_updated makes more sense as the time of the last commit/release, but I believe that means some dates should be changed

File: api/bazel/external_deps.bzl (hunk 1 above)
Comment (htuch) [reply api/bazel/external_deps.bzl:21]: This is now tracked in https://github.com/envoyproxy/envoy/pull/13582/files

File: tools/dependency/validate.py (hunk 2 above)
Comment (htuch) [reply tools/dependency/validate.py:206]: Yeah, not a ton of value in disentangling these, whereas on dataplane, we do want to be clear about what is marginally contribute by certain extensions. Will add a comment.

**Comment by moderation** [comment]
> Thanks! I personally think last_updated makes more sense as the time of the last commit/release, but I believe that means some dates should be changed

That is the plan @asraa. Harvey linked to the new PR where we switch from `last_updated` to `release_date` and make it refer to last commit / release time as you suggest.

**Review by asraa — APPROVED** [review]
Thank you!

**Pull request merged**